File: rustbridge-csharp/RustBridge.Benchmarks/BenchmarkHelper.cs
namespace RustBridge.Benchmarks;

/// <summary>
/// Helper class for finding the hello-plugin library.
/// </summary>
public static class BenchmarkHelper
{
    /// <summary>
    /// Find the hello-plugin library, searching in common locations.
    /// </summary>
    /// <returns>Full path to the library, or null if not found.</returns>
    public static string? FindHelloPlugin()
    {
        var libraryName = GetLibraryFileName("hello_plugin");
        var assemblyLocation = typeof(BenchmarkHelper).Assembly.Location;
        var assemblyDir = Path.GetDirectoryName(assemblyLocation) ?? ".";

        // Search from current directory
        var result = SearchForPlugin(Environment.CurrentDirectory, libraryName);
        if (result != null) return result;

        // Search from assembly location, walking up the directory tree
        // BenchmarkDotNet copies assemblies to deep temp paths like:
        // bin\Release\net8.0\{guid}\bin\Release\net8.0
        // So we need to walk up many levels to find the repo root
        var searchDir = assemblyDir;
        for (int i = 0; i < 15; i++)
        {
            result = SearchForPlugin(searchDir, libraryName);
            if (result != null) return result;

            var parent = Path.GetDirectoryName(searchDir);
            if (parent == null || parent == searchDir) break;
            searchDir = parent;
        }

        return null;
    }

    private static string? SearchForPlugin(string baseDir, string libraryName)
    {
        // Prefer release build for benchmarks
        var releasePath = Path.Combine(baseDir, "target", "release", libraryName);
        if (File.Exists(releasePath))
        {
            return Path.GetFullPath(releasePath);
        }

        var debugPath = Path.Combine(baseDir, "target", "debug", libraryName);
        if (File.Exists(debugPath))
        {
            return Path.GetFullPath(debugPath);
        }

        return null;
    }

    /// <summary>
    /// Get the hello-plugin library path or throw if not found.
    /// </summary>
    public static string GetHelloPluginOrThrow()
    {
        return FindHelloPlugin()
            ?? throw new InvalidOperationException(
                "hello-plugin not found. Run: cargo build --release -p hello-plugin");
    }

    private static string GetLibraryFileName(string name)
    {
        if (OperatingSystem.IsWindows()) return $"{name}.dll";
        if (OperatingSystem.IsMacOS()) return $"lib{name}.dylib";
        return $"lib{name}.so";
    }
}

File: rustbridge-java/rustbridge-benchmarks/src/jmh/java/com/rustbridge/benchmarks/BenchmarkHelper.java
package com.rustbridge.benchmarks;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Helper utilities for JMH benchmarks.
 */
public final class BenchmarkHelper {

    private BenchmarkHelper() {
        // Utility class
    }

    /**
     * Find the hello_plugin shared library.
     *
     * @return path to the plugin library
     * @throws IllegalStateException if the library cannot be found
     */
    public static Path findHelloPluginLibrary() {
        String libraryName = getLibraryName();

        // Start from current directory and search upward
        Path current = Paths.get("").toAbsolutePath();
        for (int i = 0; i < 5; i++) {
            Path releaseLib = current.resolve("target/release/" + libraryName);
            if (Files.exists(releaseLib)) {
                return releaseLib;
            }

            Path debugLib = current.resolve("target/debug/" + libraryName);
            if (Files.exists(debugLib)) {
                return debugLib;
            }

            current = current.getParent();
            if (current == null) {
                break;
            }
        }

        throw new IllegalStateException(
                "Could not find " + libraryName + ". " +
                        "Run 'cargo build -p hello-plugin --release' first."
        );
    }

    private static String getLibraryName() {
        String osName = System.getProperty("os.name").toLowerCase();
        if (osName.contains("windows")) {
            return "hello_plugin.dll";
        } else if (osName.contains("mac")) {
            return "libhello_plugin.dylib";
        } else {
            return "libhello_plugin.so";
        }
    }
}
